server/src/main/java/org/elasticsearch/index/codec/vectors/ES816BinaryQuantizedVectorsWriter.java

```diff
@@ -0,0 +1,1000 @@
+        final ByteBuffer correctionsBuffer = ByteBuffer.allocate(Float.BYTES * correctionsCount).order(ByteOrder.LITTLE_ENDIAN);
+        for (int ordinal : ordMap) {
+            float[] v = fieldData.getVectors().get(ordinal);
+            float[] corrections = scalarQuantizer.quantizeForIndex(v, vector, clusterCenter);
+            binarizedVectorData.writeBytes(vector, vector.length);
+            for (int i = 0; i < corrections.length; i++) {
+                correctionsBuffer.putFloat(corrections[i]);
+            }
+            binarizedVectorData.writeBytes(correctionsBuffer.array(), correctionsBuffer.array().length);
+            correctionsBuffer.rewind();
+        }
+    }
+
+    private void writeMeta(
+        FieldInfo field,
+        int maxDoc,
+        long vectorDataOffset,
+        long vectorDataLength,
+        float[] clusterCenter,
+        float centroidDp,
+        DocsWithFieldSet docsWithField
+    ) throws IOException {
+        meta.writeInt(field.number);
+        meta.writeInt(field.getVectorEncoding().ordinal());
+        meta.writeInt(field.getVectorSimilarityFunction().ordinal());
+        meta.writeVInt(field.getVectorDimension());
+        meta.writeVLong(vectorDataOffset);
+        meta.writeVLong(vectorDataLength);
+        int count = docsWithField.cardinality();
+        meta.writeVInt(count);
+        if (count > 0) {
+            final ByteBuffer buffer = ByteBuffer.allocate(field.getVectorDimension() * Float.BYTES).order(ByteOrder.LITTLE_ENDIAN);
+            buffer.asFloatBuffer().put(clusterCenter);
+            meta.writeBytes(buffer.array(), buffer.array().length);
+            meta.writeInt(Float.floatToIntBits(centroidDp));
+        }
+        OrdToDocDISIReaderConfiguration.writeStoredMeta(
+            DIRECT_MONOTONIC_BLOCK_SHIFT,
+            meta,
+            binarizedVectorData,
+            count,
+            maxDoc,
+            docsWithField
+        );
+    }
+
+    @Override
+    public void finish() throws IOException {
+        if (finished) {
+            throw new IllegalStateException("already finished");
+        }
+        finished = true;
+        rawVectorDelegate.finish();
+        if (meta != null) {
+            // write end of fields marker
+            meta.writeInt(-1);
+            CodecUtil.writeFooter(meta);
+        }
+        if (binarizedVectorData != null) {
+            CodecUtil.writeFooter(binarizedVectorData);
+        }
+    }
+
+    @Override
+    public void mergeOneField(FieldInfo fieldInfo, MergeState mergeState) throws IOException {
+        if (fieldInfo.getVectorEncoding().equals(VectorEncoding.FLOAT32)) {
+            final float[] centroid;
+            final float[] mergedCentroid = new float[fieldInfo.getVectorDimension()];
+            int vectorCount = mergeAndRecalculateCentroids(mergeState, fieldInfo, mergedCentroid);
+            if (fieldInfo.getVectorSimilarityFunction() == COSINE) {
```

Comment:
It seems that `mergeAndRecalculateCentroids` function always returns normalized centroid for COSINE, so we can skip an extra normalization on lines 310-312.